Applied dialyzer fixes

smeevil opened the pull request:
Hi there,

I had some Dialyzer errors in my project coming from Maxwell, so I thought to give it a look.

- I  changed the type specs of conn.ex for Atom.t, Map.t, and Stream.t, which do not exist, to their original types.
- Added type spec for error.ex.
- Commented out  unreachable code in adapter.ex, which according to Dialyzer will never be triggered.
I've been tracing the calls and I can confirm this is indeed not possible since all paths that lead to this are indeed already covered. Please double check this as well, and then remove :)

There is just one error left which is :
```elixir
lib/maxwell.ex:1: The pattern _@2 = {'error', _} can never match the type {'error',_,#{'__struct__':='Elixir.Maxwell.Conn', 'method':=atom(), 'opts':=[{_,_}], 'path':=binary(), 'private':=map(), 'query_string':=map(), 'req_body':=_, 'req_headers':=#{binary()=>binary()}, 'resp_body':=binary() | maybe_improper_list(binary() | maybe_improper_list(any(),binary() | []) | byte(),binary() | []) | map(), 'resp_headers':=#{binary()=>binary()}, 'state':='error', 'status':='nil' | non_neg_integer(), 'url':=binary()}} | #{'__struct__':='Elixir.Maxwell.Conn', 'method':=atom(), 'opts':=[{atom(),_}], 'path':=binary(), 'private':=map(), 'query_string':=map(), 'req_body':='nil', 'req_headers':=#{binary()=>binary()}, 'resp_body':=binary() | maybe_improper_list(binary() | maybe_improper_list(any(),binary() | []) | byte(),binary() | []) | map(), 'resp_headers':=#{binary()=>binary()}, 'state':='sent', 'status':=non_neg_integer(), 'url':=binary()}

```

This is coming from the macro in builder.ex. As it is a macro it's really hard for me to track down the origins of the error. Basically what it is saying is that in the code an {:error, :message} is being thrown, while that should be {:error, :message, %Maxwell.Conn{}} , I could not find exactly where this happens so maybe you can have a look at that ?

Thanks for your time :)
Gerard

smeevil committed:
Fixed type specs of conn.ex, added type spec for error.ex, removed unreachable code in adapter.ex.

secretworry commented on lib/maxwell/adapter/adapter.ex:
```diff
@@ -30,9 +30,10 @@ defmodule Maxwell.Adapter do
-          other ->
-            raise "invalid return from #{unquote(__CALLER__.module)}" <>
-                  " expected Maxwell.Conn.t or {:error, reason}, but got: #{inspect other}"
+# according to dialyzer this can never match since previous clauses completely covered the type {'error',_,#{'__struct__':='Elixir.Maxwell.Conn', 'method':=_, 'opts':=_, 'path':=_, 'query_string':=_, 'req_body':=_, 'req_headers':=_, 'state':='error', 'url':=_, _=>_}} | #{'__struct__':='Elixir.Maxwell.Conn', 'method':=_, 'opts':=_, 'path':=_, 'query_string':=_, 'req_body':='nil', 'req_headers':=_, ...
```
@smeevil Actually this branch is intended to display a friendly error for adapter developers who mistakenly returned some thing unexpected. So IMHO it's not a good idea to remove it.

smeevil replied:
I totally understand that, but the code cannot be reached according to Dialyzer. 
I'll change the pull request and keep it in though.

secretworry replied:
Sorry to force you to fight against the PLT things. :P

smeevil replied:
No problem :)

secretworry reviewed (changes requested):
Keep the last catch-all branch for more user-friendly error message

smeevil committed:
updated adapter.ex to keep catch all

Restored the catch all clause, even though this code cannot be reached according to Dialyzer.

secretworry commented on lib/maxwell/adapter/adapter.ex:
```diff
@@ -30,6 +30,7 @@ defmodule Maxwell.Adapter do
         case res do
           %Conn{} -> res
           {:error, _reason, _conn} -> res
+
```
Maybe remove this empty line to avoid this file being listed in the changed files?

smeevil committed:
Update adapter.ex

smeevil commented:
Hi,

I was wondering if there is anything left that is preventing the merge for this request ?

Gerard.

secretworry commented:
@smeevil Sorry for the delay.

zhongwencool commented:
@smeevil use 2.2.2

smeevil commented:
👍

visciang commented:
@smeevil, have a look at #66.
